engine/backends/proverif/proverif_backend.ml

```diff
@@ -166,21 +165,90 @@ module Print = struct
             in
-            string "letfun" ^^ space ^^ print#concrete_ident name
-            ^^ params_string ^/^ equals
-            ^^ nest 4 (hardline ^^ print#expr_at Item_Fn_body body ^^ dot)
+            string "letfun" ^^ space
+            ^^ align
+                 (print#concrete_ident name ^^ params_string ^^ space ^^ equals
+                ^^ hardline
+                 ^^ print#expr_at Item_Fn_body body
+                 ^^ dot)
         (* `struct` definitions are transformed into simple constructors and `reduc`s for accessing fields. *)
         | Type { name; generics; variants; is_struct } ->
             if is_struct then fun_and_reduc name variants else empty
         | _ -> empty
+
+      method! expr_let : lhs:pat -> rhs:expr -> expr fn =
+        fun ~lhs ~rhs body ->
+          string "let" ^^ space
+          ^^ iblock Fn.id (print#pat_at Expr_Let_lhs lhs)
+          ^^ space ^^ equals ^^ space
+          ^^ iblock Fn.id (print#expr_at Expr_Let_rhs rhs)
+          ^^ space ^^ string "in" ^^ hardline
+          ^^ (print#expr_at Expr_Let_body body |> group)
+
+      method! doc_construct_inductive
+          : is_record:bool ->
+            is_struct:bool ->
+            constructor:concrete_ident ->
+            base:document option ->
+            (global_ident * document) list fn =
+        fun ~is_record ~is_struct:_ ~constructor ~base:_ args ->
+          if is_record then
+            string "t_"
+            (* FIXME: How do I get at the ident from the struct definition instead? *)
+            ^^ print#concrete_ident constructor
+            ^^ iblock parens
+                 (separate_map
+                    (break 0 ^^ comma)
+                    (fun (field, body) -> iblock Fn.id body |> group)
+                    args)
+          else
+            print#concrete_ident constructor
+            ^^ iblock parens (separate_map (break 0) snd args)
+
+      method ty : Generic_printer_base.par_state -> ty fn =
+        fun ctx ty ->
+          match ty with
+          | TBool -> print#ty_bool
+          | TParam i -> print#local_ident i
+          | TApp _ -> super#ty ctx ty
+          | _ -> string "bitstring"
+
+      method! expr_app : expr -> expr list -> generic_value list fn =
+        fun f args _generic_args ->
+          let dummy_fn =
+            match List.length args with
+            | n when n < 8 -> string "dummy_fn_" ^^ PPrint.OCaml.int n
+            | _ -> string "not_supported"
+          in
+          let args =
+            separate_map
+              (comma ^^ break 1)
+              (print#expr_at Expr_App_arg >> group)
+              args
+          in
+          let f = dummy_fn |> group in
+          f ^^ iblock parens args
+
+      method! literal : Generic_printer_base.literal_ctx -> literal fn =
+        fun _ctx -> function
+          | String s -> string "no char literals in ProVerif"
+          | Char c -> string "no char literals in ProVerif"
```

Comment:
You should `raise` an error just like above, and you might need to add a `Unsupported` variant to the [enum Kind](https://github.com/hacspec/hax/blob/fea5dbcf4ab371ba6cb18e8c3f7e3742d4982dd1/frontend/diagnostics/src/lib.rs#L131)

Reply:
I could introduce an `Unsupported` variant to signal that the backend doesn't support the input, but doesn't `ExplicitRejection` already cover that? We're not in a phase anymore here, but the intention seems the same to me.

Reply:
Oh, you're right, that's a good idea yes!